### ISSUE TITLE
mergerfs fuse hidden files

### ISSUE THREAD

**Issue by rickysarraf** [trapexit/mergerfs#521]
General description
===================

So I use `mergerfs` to make a large union space on `/tmp` backed by ramfs and then followed by disk. Everything works fine with the exception that, some times, after a build, there are odd FUSE hidden files in my result.

In my case, I do Debian package builds using `pbuilder`, and my build directory is set to `/tmp/`

Expected behavior
=================
There should be no external files. To me it looks like there is a race condition wherein mergers/fuse creates internal files, which are accessed by external programs, before mergerfs/fuse can rename them to their actual names.

Actual behavior
===============
```
rrs@priyasi:/var/tmp/Debian-Build/Result$ lintian multipath-tools_0.7.8-2_amd64.changes
...snipped....
E: multipath-tools: unstripped-binary-or-object lib/multipath/.fuse_hidden00008eef000009f1

rrs@priyasi:/tmp/foo$ ar x multipath-tools_0.7.8-2_amd64.deb 
19:32 ♒♒♒   ☺ 😄    

rrs@priyasi:/tmp/foo$ tar xavf data.tar.xz 
..snipped...
./lib/multipath/.fuse_hidden00008eef000009f1
rrs@priyasi:/tmp/foo$ file ./lib/multipath/.fuse_hidden00008eef000009f1
./lib/multipath/.fuse_hidden00008eef000009f1: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=7f5cf7c5a02c87827e35ecdbb0563123c473c781, with debug_info, not stripped
19:34 ♒♒♒   ☺ 😄    
rrs@priyasi:/tmp/foo$ ldd ./lib/multipath/.fuse_hidden00008eef000009f1
        linux-vdso.so.1 (0x00007ffe83d69000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f84dbc31000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f84dbe18000)
19:35 ♒♒♒   ☺ 😄    
rrs@priyasi:/tmp/foo$ objdump -a ./lib/multipath/.fuse_hidden00008eef000009f1

./lib/multipath/.fuse_hidden00008eef000009f1:     file format elf64-x86-64
./lib/multipath/.fuse_hidden00008eef000009f1

19:35 ♒♒♒   ☺ 😄    
rrs@priyasi:/tmp/foo$ objdump -x ./lib/multipath/.fuse_hidden00008eef000009f1

./lib/multipath/.fuse_hidden00008eef000009f1:     file format elf64-x86-64
./lib/multipath/.fuse_hidden00008eef000009f1
architecture: i386:x86-64, flags 0x00000150:
HAS_SYMS, DYNAMIC, D_PAGED
start address 0x0000000000001080

Program Header:
    LOAD off    0x0000000000000000 vaddr 0x0000000000000000 paddr 0x0000000000000000 align 2**12
         filesz 0x00000000000005b0 memsz 0x00000000000005b0 flags r--
    LOAD off    0x0000000000001000 vaddr 0x0000000000001000 paddr 0x0000000000001000 align 2**12
         filesz 0x00000000000001b1 memsz 0x00000000000001b1 flags r-x
    LOAD off    0x0000000000002000 vaddr 0x0000000000002000 paddr 0x0000000000002000 align 2**12
         filesz 0x00000000000000b0 memsz 0x00000000000000b0 flags r--
    LOAD off    0x0000000000002db8 vaddr 0x0000000000003db8 paddr 0x0000000000003db8 align 2**12
         filesz 0x0000000000000250 memsz 0x0000000000000258 flags rw-
 DYNAMIC off    0x0000000000002dc8 vaddr 0x0000000000003dc8 paddr 0x0000000000003dc8 align 2**3
         filesz 0x00000000000001e0 memsz 0x00000000000001e0 flags rw-
    NOTE off    0x0000000000000238 vaddr 0x0000000000000238 paddr 0x0000000000000238 align 2**2
         filesz 0x0000000000000024 memsz 0x0000000000000024 flags r--
EH_FRAME off    0x0000000000002010 vaddr 0x0000000000002010 paddr 0x0000000000002010 align 2**2
         filesz 0x0000000000000024 memsz 0x0000000000000024 flags r--
   STACK off    0x0000000000000000 vaddr 0x0000000000000000 paddr 0x0000000000000000 align 2**4
         filesz 0x0000000000000000 memsz 0x0000000000000000 flags rw-
   RELRO off    0x0000000000002db8 vaddr 0x0000000000003db8 paddr 0x0000000000003db8 align 2**0
         filesz 0x0000000000000248 memsz 0x0000000000000248 flags r--

Dynamic Section:
  NEEDED               libc.so.6
  INIT                 0x0000000000001000
  FINI                 0x00000000000011a8
  INIT_ARRAY           0x0000000000003db8
  INIT_ARRAYSZ         0x0000000000000008
  FINI_ARRAY           0x0000000000003dc0
  FINI_ARRAYSZ         0x0000000000000008
  GNU_HASH             0x0000000000000260
  STRTAB               0x00000000000003b0
  SYMTAB               0x0000000000000290
  STRSZ                0x00000000000000ad
  SYMENT               0x0000000000000018
  PLTGOT               0x0000000000003fa8
  PLTRELSZ             0x0000000000000060
  PLTREL               0x0000000000000007
  JMPREL               0x0000000000000550
  RELA                 0x00000000000004a8
  RELASZ               0x00000000000000a8
  RELAENT              0x0000000000000018
  FLAGS                0x0000000000000008
  FLAGS_1              0x0000000000000001
  VERNEED              0x0000000000000478
  VERNEEDNUM           0x0000000000000001
  VERSYM               0x000000000000045e
  RELACOUNT            0x0000000000000003

Version References:
  required from libc.so.6:
    0x09691a75 0x00 03 GLIBC_2.2.5
    0x0d696914 0x00 02 GLIBC_2.4

Sections:
Idx Name          Size      VMA               LMA               File off  Algn
  0 .note.gnu.build-id 00000024  0000000000000238  0000000000000238  00000238  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  1 .gnu.hash     00000030  0000000000000260  0000000000000260  00000260  2**3
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  2 .dynsym       00000120  0000000000000290  0000000000000290  00000290  2**3
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  3 .dynstr       000000ad  00000000000003b0  00000000000003b0  000003b0  2**0
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  4 .gnu.version  00000018  000000000000045e  000000000000045e  0000045e  2**1
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  5 .gnu.version_r 00000030  0000000000000478  0000000000000478  00000478  2**3
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  6 .rela.dyn     000000a8  00000000000004a8  00000000000004a8  000004a8  2**3
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  7 .rela.plt     00000060  0000000000000550  0000000000000550  00000550  2**3
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  8 .init         00000017  0000000000001000  0000000000001000  00001000  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
  9 .plt          00000050  0000000000001020  0000000000001020  00001020  2**4
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
 10 .plt.got      00000008  0000000000001070  0000000000001070  00001070  2**3
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
 11 .text         00000125  0000000000001080  0000000000001080  00001080  2**4
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
 12 .fini         00000009  00000000000011a8  00000000000011a8  000011a8  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
 13 .rodata       00000010  0000000000002000  0000000000002000  00002000  2**3
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
 14 .eh_frame_hdr 00000024  0000000000002010  0000000000002010  00002010  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
 15 .eh_frame     00000078  0000000000002038  0000000000002038  00002038  2**3
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
 16 .init_array   00000008  0000000000003db8  0000000000003db8  00002db8  2**3
                  CONTENTS, ALLOC, LOAD, DATA
 17 .fini_array   00000008  0000000000003dc0  0000000000003dc0  00002dc0  2**3
                  CONTENTS, ALLOC, LOAD, DATA
 18 .dynamic      000001e0  0000000000003dc8  0000000000003dc8  00002dc8  2**3
                  CONTENTS, ALLOC, LOAD, DATA
 19 .got          00000058  0000000000003fa8  0000000000003fa8  00002fa8  2**3
                  CONTENTS, ALLOC, LOAD, DATA
 20 .data         00000008  0000000000004000  0000000000004000  00003000  2**3
                  CONTENTS, ALLOC, LOAD, DATA
 21 .bss          00000008  0000000000004008  0000000000004008  00003008  2**0
                  ALLOC
 22 .comment      0000001c  0000000000000000  0000000000000000  00003008  2**0
                  CONTENTS, READONLY
 23 .debug_aranges 00000030  0000000000000000  0000000000000000  00003024  2**0
                  CONTENTS, READONLY, DEBUGGING
 24 .debug_info   00000504  0000000000000000  0000000000000000  00003054  2**0
                  CONTENTS, READONLY, DEBUGGING
 25 .debug_abbrev 0000018b  0000000000000000  0000000000000000  00003558  2**0
                  CONTENTS, READONLY, DEBUGGING
 26 .debug_line   00000187  0000000000000000  0000000000000000  000036e3  2**0
                  CONTENTS, READONLY, DEBUGGING
 27 .debug_str    000003bf  0000000000000000  0000000000000000  0000386a  2**0
                  CONTENTS, READONLY, DEBUGGING
 28 .debug_loc    000000b7  0000000000000000  0000000000000000  00003c29  2**0
                  CONTENTS, READONLY, DEBUGGING
SYMBOL TABLE:
0000000000000238 l    d  .note.gnu.build-id     0000000000000000              .note.gnu.build-id
0000000000000260 l    d  .gnu.hash      0000000000000000              .gnu.hash
0000000000000290 l    d  .dynsym        0000000000000000              .dynsym
00000000000003b0 l    d  .dynstr        0000000000000000              .dynstr
000000000000045e l    d  .gnu.version   0000000000000000              .gnu.version
0000000000000478 l    d  .gnu.version_r 0000000000000000              .gnu.version_r
00000000000004a8 l    d  .rela.dyn      0000000000000000              .rela.dyn
0000000000000550 l    d  .rela.plt      0000000000000000              .rela.plt
0000000000001000 l    d  .init  0000000000000000              .init
0000000000001020 l    d  .plt   0000000000000000              .plt
0000000000001070 l    d  .plt.got       0000000000000000              .plt.got
0000000000001080 l    d  .text  0000000000000000              .text
00000000000011a8 l    d  .fini  0000000000000000              .fini
0000000000002000 l    d  .rodata        0000000000000000              .rodata
0000000000002010 l    d  .eh_frame_hdr  0000000000000000              .eh_frame_hdr
0000000000002038 l    d  .eh_frame      0000000000000000              .eh_frame
0000000000003db8 l    d  .init_array    0000000000000000              .init_array
0000000000003dc0 l    d  .fini_array    0000000000000000              .fini_array
0000000000003dc8 l    d  .dynamic       0000000000000000              .dynamic
0000000000003fa8 l    d  .got   0000000000000000              .got
0000000000004000 l    d  .data  0000000000000000              .data
0000000000004008 l    d  .bss   0000000000000000              .bss
0000000000000000 l    d  .comment       0000000000000000              .comment
0000000000000000 l    d  .debug_aranges 0000000000000000              .debug_aranges
0000000000000000 l    d  .debug_info    0000000000000000              .debug_info
0000000000000000 l    d  .debug_abbrev  0000000000000000              .debug_abbrev
0000000000000000 l    d  .debug_line    0000000000000000              .debug_line
0000000000000000 l    d  .debug_str     0000000000000000              .debug_str
0000000000000000 l    d  .debug_loc     0000000000000000              .debug_loc
0000000000000000 l    df *ABS*  0000000000000000              crtstuff.c
0000000000001080 l     F .text  0000000000000000              deregister_tm_clones
00000000000010b0 l     F .text  0000000000000000              register_tm_clones
00000000000010f0 l     F .text  0000000000000000              __do_global_dtors_aux
0000000000004008 l     O .bss   0000000000000001              completed.7389
0000000000003dc0 l     O .fini_array    0000000000000000              __do_global_dtors_aux_fini_array_entry
0000000000001130 l     F .text  0000000000000000              frame_dummy
0000000000003db8 l     O .init_array    0000000000000000              __frame_dummy_init_array_entry
0000000000000000 l    df *ABS*  0000000000000000              random.c
0000000000000000 l    df *ABS*  0000000000000000              crtstuff.c
00000000000020ac l     O .eh_frame      0000000000000000              __FRAME_END__
0000000000000000 l    df *ABS*  0000000000000000              
0000000000004000 l     O .data  0000000000000000              __dso_handle
0000000000003dc8 l     O .dynamic       0000000000000000              _DYNAMIC
0000000000002010 l       .eh_frame_hdr  0000000000000000              __GNU_EH_FRAME_HDR
0000000000004008 l     O .data  0000000000000000              __TMC_END__
0000000000003fa8 l     O .got   0000000000000000              _GLOBAL_OFFSET_TABLE_
0000000000000000  w      *UND*  0000000000000000              _ITM_deregisterTMCloneTable
00000000000011a8 g     F .fini  0000000000000000              _fini
0000000000000000       F *UND*  0000000000000000              __stack_chk_fail@@GLIBC_2.4
0000000000000000       F *UND*  0000000000000000              gettimeofday@@GLIBC_2.2.5
0000000000000000       F *UND*  0000000000000000              srand@@GLIBC_2.2.5
0000000000000000  w      *UND*  0000000000000000              __gmon_start__
0000000000000000  w      *UND*  0000000000000000              _ITM_registerTMCloneTable
0000000000001140 g     F .text  0000000000000065              getprio
0000000000000000  w    F *UND*  0000000000000000              __cxa_finalize@@GLIBC_2.2.5
0000000000001000 g     F .init  0000000000000000              _init
0000000000000000       F *UND*  0000000000000000              rand@@GLIBC_2.2.5


19:35 ♒♒♒   ☺ 😄    
```

**Precise** steps to reproduce the behavior
===========================================

> Perhaps the best is to setup a `/tmp/` setup with mergerfs
> Then build packages in `/tmp/` with Debian's `pbuilder`

System information
==================
Please provide as much of the following information as possible:

* [x] mergerfs version: `mergerfs -V`
rrs@priyasi:/tmp/foo$ mergerfs -V
mergerfs version: 2.24.2
FUSE library version: 2.9.8-mergerfs
fusermount version: 2.9.8
using FUSE kernel interface version 7.19
19:38 ♒♒♒    ☹ 😟=> 1  

Why is this command giving an output of exit status `1`

* [x] mergerfs settings: `cat /etc/fstab` or the command line arguments
```
# mergerfs
tmpfs   /ram-tmp        tmpfs   defaults        0       0
/ram-tmp:/disk-tmp      /tmp    fuse.mergerfs   x-systemd.requires=/ram-tmp,defaults,allow_other,use_ino,nonempty,minfreespace=1G,category.create=all    0       0
```

* [x] Linux version: `uname -a`
```
rrs@priyasi:/tmp/foo$ uname -a
Linux priyasi 4.18.0-2-amd64 #1 SMP Debian 4.18.10-2 (2018-10-07) x86_64 GNU/Linux
```

Debian Unstable

**Comment by trapexit** [comment]
Those are removed files which are still open. It's a standard FUSE behavior. You should add `hard_remove` to your settings.

**Comment by rickysarraf** [comment]
That option `hard_remove`, when used over a merged fs of `/tmp`, breaks just too many things. Which would mean that whatever it does internally, it is not doing it transparent to its userspace.

So, all applications that depends on TMPDIR for some task, get broken. In my case, `apt` and `pbuilder` are 2 examples that immediately broke.

From `pbuilder`:
```
dpkg-source: warning: extracting unsigned source package (bpfcc_0.7.0-1.dsc)
dpkg-source: info: extracting bpfcc in bpfcc-0.7.0
dpkg-source: info: unpacking bpfcc_0.7.0.orig.tar.gz
dpkg-source: info: unpacking bpfcc_0.7.0-1.debian.tar.xz
dpkg-source: info: using patch list from debian/patches/series
dpkg-source: info: applying fix-install-path.patch
dpkg-source: info: applying fix-uint128-build-failure.patch
I: using fakeroot in build.
I: Installing the build-deps
I: user script /tmp//6171/tmp/hooks/D10-man-db starting
I: Preseed man-db/auto-update to false
I: user script /tmp//6171/tmp/hooks/D10-man-db finished
I: user script /tmp//6171/tmp/hooks/D70results starting
No '/var/tmp/Debian-Build/pbuilder-deps/Packages' file found. Regular build. So no local package db update
I: user script /tmp//6171/tmp/hooks/D70results finished
 -> Attempting to satisfy build-dependencies
 -> Creating pbuilder-satisfydepends-dummy package
cat: -: No such file or directory
E: pbuilder-satisfydepends failed.
I: unmounting /var/cache/apt/archives/ filesystem
I: unmounting dev/ptmx filesystem
I: unmounting dev/pts filesystem
```

And `apt`:
```
rrs@priyasi:~$ sudo apt update
[sudo] password for rrs: 
Hit:1 http://security.debian.org/debian-security testing/updates InRelease
Hit:2 http://deb.debian.org/debian testing InRelease
Hit:3 http://deb.debian.org/debian unstable InRelease
Hit:4 http://deb.debian.org/debian experimental InRelease
Reading package lists... Error!
E: Unable to determine file size for fd 9 - fstat (2: No such file or directory)
E: The repository 'http://security.debian.org/debian-security testing/updates InRelease' provides only weak security information.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: Unable to determine file size for fd 9 - fstat (2: No such file or directory)
E: The repository 'http://deb.debian.org/debian testing InRelease' provides only weak security information.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: Unable to determine file size for fd 9 - fstat (2: No such file or directory)
E: The repository 'http://deb.debian.org/debian unstable InRelease' provides only weak security information.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: Unable to determine file size for fd 9 - fstat (2: No such file or directory)
E: The repository 'http://deb.debian.org/debian experimental InRelease' provides only weak security information.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: Unable to determine file size for fd 4 - fstat (2: No such file or directory)
E: The package lists or status file could not be parsed or opened.
19:20 ♒♒♒    ☹ 😟=> 100  
```

**Comment by trapexit** [comment]
Without knowing what those apps are doing I can't really comment. I'll need a strace of both the app and mergerfs. If you can do so great otherwise I'll try to reproduce it.

The hidden files are a feature. Not a bug. It's how fuse is handling unlinked files still in use. hard_remove disables that and would break something using the unlinked file after it was unlinked. However, according to the docs the way I have things setup it shouldn't in fact have that problem.

**Comment by trapexit** [comment]
Preferably I'd not need to mess with my system's /tmp. Is there a way to point apt elsewhere?

**Comment by trapexit** [comment]
"what broke"?

Are you saying mergerfs' behavior changed?

**Comment by trapexit** [comment]
I'm not too familiar with pbuilder but I imagine the build is adding whatever files exist in a directory? I've ideas on how better to solve this issue but for now could you try building and using the "hidden_files" branch and don't use hard_remove.

**Comment by rickysarraf** [comment]
Thanks you. I have built and am testing now. I will post the results here.

> "what broke"?
> 
> Are you saying mergerfs' behavior changed?

I meant the applications that rely on `/tmp` for normal file system behavior.

**Comment by rickysarraf** [comment]
So today, I tried building using the `/tmp` mount point, which is `mergerfs` based. Co-incidentally, it is `mergerfs` that I need to build to upload a newer version to Debian. :-)

```
rrs@priyasi:~/rrs-home/Community/Packaging/mergerfs (master)$ export TMPDIR=/tmp; gbp buildpackage --git-ignore-new 
dh clean --parallel --with=autoreconf
   dh_auto_clean -O--parallel
        make V=1 -j4 distclean
make[1]: Entering directory '/home/rrs/NoBackup/Community/Packaging/mergerfs'
Makefile:40: "pandoc does not appear available: manpage won't be buildable"
Makefile:77: "xattr not available: disabling"
/bin/rm -rf rpmbuild
cd libfuse && \
/bin/mkdir -p m4 && \
autoreconf --force --install && \
        ./configure --enable-lib --disable-util --disable-example
libtoolize: putting auxiliary files in '.'.
libtoolize: copying file './ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
configure.ac:10: installing './compile'
configure.ac:6: installing './missing'
example/Makefile.am: installing './depcomp'
cannot remove directory for /tmp/arJaT7rh/m4-qrL7gi: Directory not empty at /usr/share/autoconf/Autom4te/General.pm line 196.
cannot remove directory for /tmp/arJaT7rh/m4-0HcvYJ: Directory not empty at /usr/share/autoconf/Autom4te/General.pm line 196.
cannot remove directory for /tmp/arJaT7rh/m4-QzlO04: Directory not empty at /usr/share/autoconf/Autom4te/General.pm line 196.
cannot remove directory for /tmp/arJaT7rh/m4-1QH1R6: Directory not empty at /usr/share/autoconf/Autom4te/General.pm line 196.
make[1]: *** [Makefile:215: libfuse_Makefile] Error 1
make[1]: Leaving directory '/home/rrs/NoBackup/Community/Packaging/mergerfs'
dh_auto_clean: make V=1 -j4 distclean returned exit code 2
make: *** [debian/rules:15: clean] Error 2
gbp:error: 'fakeroot debian/rules clean' failed: it exited with 2
12:25 ♒♒♒    ☹ 😟=> 1  
```
I have verified already. If I set TMPDIR to a non-`mergerfs` based path, the build proceeds further.

**Comment by rickysarraf** [comment]
Here's the strace log during the above mentioned operation.
[mergerfs.strace.log](https://github.com/trapexit/mergerfs/files/2592565/mergerfs.strace.log)

**Comment by trapexit** [comment]
This is with the hidden file branch?

**Comment by rickysarraf** [comment]
> This is with the hidden file branch?

Yes. I am still on the `hidden` branch.

**Comment by rickysarraf** [comment]
> I suspect it's trying to delete directories and they aren't empty because of the hidden files. Is there no way to blacklist files to pick up? This is a general libfuse issue and would require some significant changes that would only work by changing the library upstream (or changing my vendored version but then it won't work on system versions.)

Yes. I am sure that must be the case. Because `gbp` is calling the `clean` target.

**Comment by trapexit** [comment]
I've some ideas but they are a little invasive. I'll need to test them out.

**Comment by trapexit** [comment]
Yes, choosing to move it elsewhere would work but it could interfere (at least logically) with policies though this could be a special case. A problem is that I need to audit to find out exactly which functions might be called. Ideally I'd not rename the files at all and keep an open file descriptor till the file is fully closed but I'm not sure that's possible given how fuse handles the situation. I could change libfuse (cleaner) or I'd need to catch every request that takes a path and check if the filename is .fuse_hidden*. There really needed to be a `hide` function rather than overloading `rename`. But to do that means mergerfs wouldn't build against standard libfuse or it'd need two versions. I can't simply move to root dir because you might not have permission to do so. I suppose I could check in `rename` if it's a "hide" then switch to root and rename to root but that's kindof ugly. Also, edge case but, what if you're working in the root dir?

.mergerfs is a fake file. It doesn't actually exist but in memory/code. These are real objects on the underlying filesystem which need to be stored somewhere.

**Comment by ParadingLunatic** [comment]
I JUST ran across this issue on Wednesday.  I'm using Nextcloud with its data folder over NFS mount that is exported from a MergerFS storage.  When it went to run its cleanup it was constantly erroring on the rmdir on most of the folders due to the .fuse-hidden files.  I just told it to retry and it eventually finished, but only after having to click retry what felt like hundreds of times.  Haven't added the hard_remove option yet.

**Comment by trapexit** [comment]
I've a change / fix in the works. It's extremely invasive but it should cover most situations. Both libfuse and the FUSE kernel driver are at fault. Can't do much about the latter but given I've vendored the library anyway I can make otherwise breaking changes. I should have something to test this week.

**Comment by trapexit** [comment]
@rickysarraf @ParadingLunatic 

Can you try out the [hidden](https://github.com/trapexit/mergerfs/tree/hidden) branch?

I completely ripped out the `.fuse_hidden*` behavior and on unlink / rename will open the file again and then perform the unlink / rename. That file descriptor is kept open till the release comes in (indicating all those who had it open close it). When a `getattr` or `setattr` comes in it will use that extra file handle. 

Only works on files right now. Would take a bit more research and hacking to get it to work on directories but that's a less common behavior. 

Ideally I'd rewrite mergerfs to use the low level API and handle this more explicitly but given I'm using the high level API... hacking the HLAPI is the easiest route.

**Comment by rickysarraf** [comment]
@trapexit I have built your `hidden` branch here, locally. I will test it a couple of days and share my findings. Thanks for working on this corner case.

**Comment by trapexit** [comment]
No problem. It's a better behavior even if its hacky. This lead me to rediscover the fact that the kernel is not including the file handle with all calls as one would expect. Doesn't help us now but I'm looking into what can be done on the kernel side to help.

**Comment by trapexit** [comment]
Any updates?

**Comment by rickysarraf** [comment]
> Any updates?

Sorry for the delay. Fridays can be unpredictable :-)

For me, so far, it looks all good. There are no more stray hidden files.

**Comment by trapexit** [comment]
My main concern is stability and correctness. I completely ripped out the hidden file stuff so I know that'd be gone :)

I've written a few tests to check this stuff but need to create a more thorough framework to orchestrate the running of said tests.

**Comment by trapexit** [comment]
pgrep

**Comment by rickysarraf** [comment]
So... Today I ran a build job to build `ruby2.5`. It runs a large number of tests. I think this must be a fuse corner case but I thought of reporting it to you. I am not sure how persistently reproducible is it. I will try to reboot the machine some time tomorrow and check if the issue is seen again.

The build job basically builds the entire package in `/tmp`, including execution of its tests. Steps to reproduce should be something like:

* cd /tmp
* apt source ruby2.5
* cd to ruby2.5 source folder
* `dpkg-buildpackage -uc -us`

```
[157200.661402] INFO: task ruby2.5:22867 blocked for more than 120 seconds.
[157200.661405]       Tainted: G     U  W         4.19.0-4-amd64 #1 Debian 4.19.28-2
[157200.661406] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[157200.661408] ruby2.5         D    0 22867  20154 0x00000004
[157200.661410] Call Trace:
[157200.661416]  ? __schedule+0x2a2/0x870
[157200.661417]  schedule+0x28/0x80
[157200.661422]  request_wait_answer+0x12b/0x210 [fuse]
[157200.661425]  ? finish_wait+0x80/0x80
[157200.661428]  __fuse_request_send+0x64/0x90 [fuse]
[157200.661431]  fuse_simple_request+0xcd/0x190 [fuse]
[157200.661433]  fuse_setlk+0x160/0x1b0 [fuse]
[157200.661437]  __x64_sys_flock+0xa4/0xf0
[157200.661439]  do_syscall_64+0x53/0x100
[157200.661441]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[157200.661443] RIP: 0033:0x7f9605280cf7
[157200.661446] Code: Bad RIP value.
[157200.661447] RSP: 002b:00007ffc246a1a98 EFLAGS: 00000206 ORIG_RAX: 0000000000000049
[157200.661448] RAX: ffffffffffffffda RBX: 000055aa46afce70 RCX: 00007f9605280cf7
[157200.661449] RDX: 000055aa46afc960 RSI: 0000000000000001 RDI: 0000000000000007
[157200.661450] RBP: 0000000000000001 R08: 0000000000000000 R09: 0000000000000000
[157200.661450] R10: 000055aa46e7bf30 R11: 0000000000000206 R12: 00007f960565d2b0
[157200.661451] R13: 00007ffc246a1c58 R14: 0000000000000001 R15: 000055aa46b3d148
[157321.451035] INFO: task ruby2.5:22867 blocked for more than 120 seconds.
[157321.451043]       Tainted: G     U  W         4.19.0-4-amd64 #1 Debian 4.19.28-2
[157321.451044] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[157321.451047] ruby2.5         D    0 22867  20154 0x00000004
[157321.451049] Call Trace:
[157321.451060]  ? __schedule+0x2a2/0x870
[157321.451063]  schedule+0x28/0x80
[157321.451076]  request_wait_answer+0x12b/0x210 [fuse]
[157321.451088]  ? finish_wait+0x80/0x80
[157321.451090]  __fuse_request_send+0x64/0x90 [fuse]
[157321.451092]  fuse_simple_request+0xcd/0x190 [fuse]
[157321.451096]  fuse_setlk+0x160/0x1b0 [fuse]
[157321.451104]  __x64_sys_flock+0xa4/0xf0
[157321.451114]  do_syscall_64+0x53/0x100
[157321.451118]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[157321.451121] RIP: 0033:0x7f9605280cf7
[157321.451130] Code: Bad RIP value.
[157321.451132] RSP: 002b:00007ffc246a1a98 EFLAGS: 00000206 ORIG_RAX: 0000000000000049
[157321.451134] RAX: ffffffffffffffda RBX: 000055aa46afce70 RCX: 00007f9605280cf7
[157321.451134] RDX: 000055aa46afc960 RSI: 0000000000000001 RDI: 0000000000000007
[157321.451135] RBP: 0000000000000001 R08: 0000000000000000 R09: 0000000000000000
[157321.451135] R10: 000055aa46e7bf30 R11: 0000000000000206 R12: 00007f960565d2b0
[157321.451136] R13: 00007ffc246a1c58 R14: 0000000000000001 R15: 000055aa46b3d148
[157442.316015] INFO: task ruby2.5:22867 blocked for more than 120 seconds.
[157442.316026]       Tainted: G     U  W         4.19.0-4-amd64 #1 Debian 4.19.28-2
[157442.316027] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[157442.316030] ruby2.5         D    0 22867  20154 0x00000004
[157442.316035] Call Trace:
[157442.316052]  ? __schedule+0x2a2/0x870
[157442.316054]  schedule+0x28/0x80
[157442.316069]  request_wait_answer+0x12b/0x210 [fuse]
[157442.316082]  ? finish_wait+0x80/0x80
[157442.316085]  __fuse_request_send+0x64/0x90 [fuse]
[157442.316088]  fuse_simple_request+0xcd/0x190 [fuse]
[157442.316092]  fuse_setlk+0x160/0x1b0 [fuse]
[157442.316102]  __x64_sys_flock+0xa4/0xf0
[157442.316113]  do_syscall_64+0x53/0x100
[157442.316119]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[157442.316122] RIP: 0033:0x7f9605280cf7
[157442.316134] Code: Bad RIP value.
[157442.316135] RSP: 002b:00007ffc246a1a98 EFLAGS: 00000206 ORIG_RAX: 0000000000000049
[157442.316138] RAX: ffffffffffffffda RBX: 000055aa46afce70 RCX: 00007f9605280cf7
[157442.316139] RDX: 000055aa46afc960 RSI: 0000000000000001 RDI: 0000000000000007
[157442.316139] RBP: 0000000000000001 R08: 0000000000000000 R09: 0000000000000000
[157442.316141] R10: 000055aa46e7bf30 R11: 0000000000000206 R12: 00007f960565d2b0
[157442.316141] R13: 00007ffc246a1c58 R14: 0000000000000001 R15: 000055aa46b3d148
[157563.185798] INFO: task ruby2.5:22867 blocked for more than 120 seconds.
[157563.185804]       Tainted: G     U  W         4.19.0-4-amd64 #1 Debian 4.19.28-2
[157563.185805] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[157563.185807] ruby2.5         D    0 22867  20154 0x00000004
[157563.185810] Call Trace:
[157563.185817]  ? __schedule+0x2a2/0x870
[157563.185819]  schedule+0x28/0x80
[157563.185825]  request_wait_answer+0x12b/0x210 [fuse]
[157563.185830]  ? finish_wait+0x80/0x80
[157563.185833]  __fuse_request_send+0x64/0x90 [fuse]
[157563.185836]  fuse_simple_request+0xcd/0x190 [fuse]
[157563.185840]  fuse_setlk+0x160/0x1b0 [fuse]
[157563.185845]  __x64_sys_flock+0xa4/0xf0
[157563.185848]  do_syscall_64+0x53/0x100
[157563.185851]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[157563.185853] RIP: 0033:0x7f9605280cf7
[157563.185858] Code: Bad RIP value.
[157563.185859] RSP: 002b:00007ffc246a1a98 EFLAGS: 00000206 ORIG_RAX: 0000000000000049
[157563.185861] RAX: ffffffffffffffda RBX: 000055aa46afce70 RCX: 00007f9605280cf7
[157563.185862] RDX: 000055aa46afc960 RSI: 0000000000000001 RDI: 0000000000000007
[157563.185863] RBP: 0000000000000001 R08: 0000000000000000 R09: 0000000000000000
[157563.185864] R10: 000055aa46e7bf30 R11: 0000000000000206 R12: 00007f960565d2b0
[157563.185865] R13: 00007ffc246a1c58 R14: 0000000000000001 R15: 000055aa46b3d148
[157683.970659] INFO: task ruby2.5:22867 blocked for more than 120 seconds.
[157683.970665]       Tainted: G     U  W         4.19.0-4-amd64 #1 Debian 4.19.28-2
[157683.970667] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[157683.970669] ruby2.5         D    0 22867  20154 0x00000004
[157683.970672] Call Trace:
[157683.970680]  ? __schedule+0x2a2/0x870
[157683.970683]  schedule+0x28/0x80
[157683.970690]  request_wait_answer+0x12b/0x210 [fuse]
[157683.970696]  ? finish_wait+0x80/0x80
[157683.970700]  __fuse_request_send+0x64/0x90 [fuse]
[157683.970704]  fuse_simple_request+0xcd/0x190 [fuse]
[157683.970709]  fuse_setlk+0x160/0x1b0 [fuse]
[157683.970714]  __x64_sys_flock+0xa4/0xf0
[157683.970718]  do_syscall_64+0x53/0x100
[157683.970722]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[157683.970724] RIP: 0033:0x7f9605280cf7
[157683.970730] Code: Bad RIP value.
[157683.970732] RSP: 002b:00007ffc246a1a98 EFLAGS: 00000206 ORIG_RAX: 0000000000000049
[157683.970734] RAX: ffffffffffffffda RBX: 000055aa46afce70 RCX: 00007f9605280cf7
[157683.970735] RDX: 000055aa46afc960 RSI: 0000000000000001 RDI: 0000000000000007
[157683.970736] RBP: 0000000000000001 R08: 0000000000000000 R09: 0000000000000000
[157683.970737] R10: 000055aa46e7bf30 R11: 0000000000000206 R12: 00007f960565d2b0
[157683.970738] R13: 00007ffc246a1c58 R14: 0000000000000001 R15: 000055aa46b3d148
[157804.736436] INFO: task ruby2.5:22867 blocked for more than 120 seconds.
[157804.736448]       Tainted: G     U  W         4.19.0-4-amd64 #1 Debian 4.19.28-2
[157804.736452] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[157804.736457] ruby2.5         D    0 22867  20154 0x00000004
[157804.736463] Call Trace:
[157804.736480]  ? __schedule+0x2a2/0x870
[157804.736485]  schedule+0x28/0x80
[157804.736500]  request_wait_answer+0x12b/0x210 [fuse]
[157804.736512]  ? finish_wait+0x80/0x80
[157804.736521]  __fuse_request_send+0x64/0x90 [fuse]
[157804.736531]  fuse_simple_request+0xcd/0x190 [fuse]
[157804.736541]  fuse_setlk+0x160/0x1b0 [fuse]
[157804.736551]  __x64_sys_flock+0xa4/0xf0
[157804.736561]  do_syscall_64+0x53/0x100
[157804.736568]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[157804.736573] RIP: 0033:0x7f9605280cf7
[157804.736585] Code: Bad RIP value.
[157804.736588] RSP: 002b:00007ffc246a1a98 EFLAGS: 00000206 ORIG_RAX: 0000000000000049
[157804.736594] RAX: ffffffffffffffda RBX: 000055aa46afce70 RCX: 00007f9605280cf7
[157804.736596] RDX: 000055aa46afc960 RSI: 0000000000000001 RDI: 0000000000000007
[157804.736599] RBP: 0000000000000001 R08: 0000000000000000 R09: 0000000000000000
[157804.736601] R10: 000055aa46e7bf30 R11: 0000000000000206 R12: 00007f960565d2b0
[157804.736603] R13: 00007ffc246a1c58 R14: 0000000000000001 R15: 000055aa46b3d148
[157925.546967] INFO: task ruby2.5:22867 blocked for more than 120 seconds.
[157925.546973]       Tainted: G     U  W         4.19.0-4-amd64 #1 Debian 4.19.28-2
[157925.546974] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[157925.546977] ruby2.5         D    0 22867  20154 0x00000004
[157925.546980] Call Trace:
[157925.546988]  ? __schedule+0x2a2/0x870
[157925.546990]  schedule+0x28/0x80
[157925.546998]  request_wait_answer+0x12b/0x210 [fuse]
[157925.547003]  ? finish_wait+0x80/0x80
[157925.547007]  __fuse_request_send+0x64/0x90 [fuse]
[157925.547011]  fuse_simple_request+0xcd/0x190 [fuse]
[157925.547016]  fuse_setlk+0x160/0x1b0 [fuse]
[157925.547021]  __x64_sys_flock+0xa4/0xf0
[157925.547025]  do_syscall_64+0x53/0x100
[157925.547029]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[157925.547031] RIP: 0033:0x7f9605280cf7
[157925.547037] Code: Bad RIP value.
[157925.547038] RSP: 002b:00007ffc246a1a98 EFLAGS: 00000206 ORIG_RAX: 0000000000000049
[157925.547041] RAX: ffffffffffffffda RBX: 000055aa46afce70 RCX: 00007f9605280cf7
[157925.547042] RDX: 000055aa46afc960 RSI: 0000000000000001 RDI: 0000000000000007
[157925.547043] RBP: 0000000000000001 R08: 0000000000000000 R09: 0000000000000000
[157925.547044] R10: 000055aa46e7bf30 R11: 0000000000000206 R12: 00007f960565d2b0
[157925.547045] R13: 00007ffc246a1c58 R14: 0000000000000001 R15: 000055aa46b3d148
[158046.388316] INFO: task ruby2.5:22867 blocked for more than 120 seconds.
[158046.388327]       Tainted: G     U  W         4.19.0-4-amd64 #1 Debian 4.19.28-2
[158046.388331] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[158046.388336] ruby2.5         D    0 22867  20154 0x00000004
[158046.388342] Call Trace:
[158046.388357]  ? __schedule+0x2a2/0x870
[158046.388362]  schedule+0x28/0x80
[158046.388376]  request_wait_answer+0x12b/0x210 [fuse]
[158046.388387]  ? finish_wait+0x80/0x80
[158046.388395]  __fuse_request_send+0x64/0x90 [fuse]
[158046.388405]  fuse_simple_request+0xcd/0x190 [fuse]
[158046.388416]  fuse_setlk+0x160/0x1b0 [fuse]
[158046.388426]  __x64_sys_flock+0xa4/0xf0
[158046.388434]  do_syscall_64+0x53/0x100
[158046.388441]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[158046.388446] RIP: 0033:0x7f9605280cf7
[158046.388457] Code: Bad RIP value.
[158046.388460] RSP: 002b:00007ffc246a1a98 EFLAGS: 00000206 ORIG_RAX: 0000000000000049
[158046.388465] RAX: ffffffffffffffda RBX: 000055aa46afce70 RCX: 00007f9605280cf7
[158046.388468] RDX: 000055aa46afc960 RSI: 0000000000000001 RDI: 0000000000000007
[158046.388470] RBP: 0000000000000001 R08: 0000000000000000 R09: 0000000000000000
[158046.388473] R10: 000055aa46e7bf30 R11: 0000000000000206 R12: 00007f960565d2b0
[158046.388475] R13: 00007ffc246a1c58 R14: 0000000000000001 R15: 000055aa46b3d148
[158167.260680] INFO: task ruby2.5:22867 blocked for more than 120 seconds.
[158167.260692]       Tainted: G     U  W         4.19.0-4-amd64 #1 Debian 4.19.28-2
[158167.260696] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[158167.260701] ruby2.5         D    0 22867  20154 0x00000004
[158167.260707] Call Trace:
[158167.260724]  ? __schedule+0x2a2/0x870
[158167.260730]  schedule+0x28/0x80
[158167.260745]  request_wait_answer+0x12b/0x210 [fuse]
[158167.260757]  ? finish_wait+0x80/0x80
[158167.260770]  __fuse_request_send+0x64/0x90 [fuse]
[158167.260783]  fuse_simple_request+0xcd/0x190 [fuse]
[158167.260799]  fuse_setlk+0x160/0x1b0 [fuse]
[158167.260812]  __x64_sys_flock+0xa4/0xf0
[158167.260821]  do_syscall_64+0x53/0x100
[158167.260828]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[158167.260833] RIP: 0033:0x7f9605280cf7
[158167.260845] Code: Bad RIP value.
[158167.260847] RSP: 002b:00007ffc246a1a98 EFLAGS: 00000206 ORIG_RAX: 0000000000000049
[158167.260853] RAX: ffffffffffffffda RBX: 000055aa46afce70 RCX: 00007f9605280cf7
[158167.260855] RDX: 000055aa46afc960 RSI: 0000000000000001 RDI: 0000000000000007
[158167.260858] RBP: 0000000000000001 R08: 0000000000000000 R09: 0000000000000000
[158167.260860] R10: 000055aa46e7bf30 R11: 0000000000000206 R12: 00007f960565d2b0
[158167.260862] R13: 00007ffc246a1c58 R14: 0000000000000001 R15: 000055aa46b3d148
[158264.256636] CPU4: Core temperature above threshold, cpu clock throttled (total events = 10231)
[158264.256637] CPU0: Core temperature above threshold, cpu clock throttled (total events = 10231)
[158264.256638] CPU5: Package temperature above threshold, cpu clock throttled (total events = 20067)
[158264.256639] CPU1: Package temperature above threshold, cpu clock throttled (total events = 20067)
[158264.256641] CPU0: Package temperature above threshold, cpu clock throttled (total events = 20067)
[158264.256642] CPU4: Package temperature above threshold, cpu clock throttled (total events = 20067)
[158264.256679] CPU7: Package temperature above threshold, cpu clock throttled (total events = 20067)
[158264.256680] CPU6: Package temperature above threshold, cpu clock throttled (total events = 20067)
[158264.256681] CPU3: Package temperature above threshold, cpu clock throttled (total events = 20067)
[158264.256682] CPU2: Package temperature above threshold, cpu clock throttled (total events = 20067)
[158264.257763] CPU0: Core temperature/speed normal
[158264.257763] CPU4: Core temperature/speed normal
[158264.257764] CPU1: Package temperature/speed normal
[158264.257765] CPU6: Package temperature/speed normal
[158264.257766] CPU2: Package temperature/speed normal
[158264.257767] CPU3: Package temperature/speed normal
[158264.257768] CPU7: Package temperature/speed normal
[158264.257768] CPU5: Package temperature/speed normal
[158264.257769] CPU4: Package temperature/speed normal
[158264.257770] CPU0: Package temperature/speed normal
[158288.109151] INFO: task ruby2.5:22867 blocked for more than 120 seconds.
[158288.109188]       Tainted: G     U  W         4.19.0-4-amd64 #1 Debian 4.19.28-2
[158288.109195] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
[158288.109202] ruby2.5         D    0 22867  20154 0x00000004
[158288.109212] Call Trace:
[158288.109234]  ? __schedule+0x2a2/0x870
[158288.109243]  schedule+0x28/0x80
[158288.109264]  request_wait_answer+0x12b/0x210 [fuse]
[158288.109281]  ? finish_wait+0x80/0x80
[158288.109295]  __fuse_request_send+0x64/0x90 [fuse]
[158288.109310]  fuse_simple_request+0xcd/0x190 [fuse]
[158288.109327]  fuse_setlk+0x160/0x1b0 [fuse]
[158288.109343]  __x64_sys_flock+0xa4/0xf0
[158288.109356]  do_syscall_64+0x53/0x100
[158288.109367]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
[158288.109374] RIP: 0033:0x7f9605280cf7
[158288.109391] Code: Bad RIP value.
[158288.109396] RSP: 002b:00007ffc246a1a98 EFLAGS: 00000206 ORIG_RAX: 0000000000000049
[158288.109405] RAX: ffffffffffffffda RBX: 000055aa46afce70 RCX: 00007f9605280cf7
[158288.109409] RDX: 000055aa46afc960 RSI: 0000000000000001 RDI: 0000000000000007
[158288.109412] RBP: 0000000000000001 R08: 0000000000000000 R09: 0000000000000000
[158288.109416] R10: 000055aa46e7bf30 R11: 0000000000000206 R12: 00007f960565d2b0
[158288.109420] R13: 00007ffc246a1c58 R14: 0000000000000001 R15: 000055aa46b3d148
[158401.298498] systemd-coredump[25596]: Failed to get EXE, ignoring: No such process
```

**Comment by trapexit** [comment]
It's possible this is an underlying device issue but possibly a kernel problem as well. I imagine its not a mergerfs issue but if you wish to document / track it here that's fine but please create a new ticket.
